plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs

```diff
@@ -10,70 +10,105 @@
+type CekMachineException = MachineException Void
+
 -- | A 'Value' packed together with the environment it's defined in.
 data Closure = Closure
-    { _closureEnvironment :: Environment
-    , _closureValue       :: Plain Value
+    { _closureVarEnv :: VarEnv
+    , _closureValue  :: Plain Value
     }
 
--- | Environments used by the CEK machine.
+-- | Variable environments used by the CEK machine.
 -- Each row is a mapping from the 'Unique' representing a variable to a 'Closure'.
-newtype Environment = Environment (IntMap Closure)
+newtype VarEnv = VarEnv (IntMap Closure)
+
+-- | The environment the CEK machine runs in.
+data CekEnv = CekEnv
+    { _cekEnvDbnms  :: DynamicBuiltinNameMeanings
+    , _cekEnvVarEnv :: VarEnv
+    }
+
+-- | The monad the CEK machine runs in.
+type CekM = ReaderT CekEnv (Either CekMachineException)
 
 data Frame
-    = FrameApplyFun Environment (Plain Value)    -- ^ @[V _]@
-    | FrameApplyArg Environment (Plain Term)     -- ^ @[_ N]@
+    = FrameApplyFun VarEnv (Plain Value)         -- ^ @[V _]@
+    | FrameApplyArg VarEnv (Plain Term)          -- ^ @[_ N]@
     | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
     | FrameUnwrap                                -- ^ @(unwrap _)@
     | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
 
 type Context = [Frame]
 
+-- | Get the current 'VarEnv'.
+getVarEnv :: CekM VarEnv
+getVarEnv = asks _cekEnvVarEnv
+
+-- | Set a new 'VarEnv' and proceed.
+withVarEnv :: VarEnv -> CekM a -> CekM a
+withVarEnv env = local $ \cekEnv -> cekEnv { _cekEnvVarEnv = env }
+
 -- | Extend an environment with a variable name, the value the variable stands for
 -- and the environment the value is defined in.
-extendEnvironment :: Name () -> Plain Value -> Environment -> Environment -> Environment
-extendEnvironment argName arg argEnv (Environment oldEnv) =
-    Environment $ IntMap.insert (unUnique $ nameUnique argName) (Closure argEnv arg) oldEnv
-
--- | Look up a name in an environment.
-lookupName :: Name () -> Environment -> Maybe Closure
-lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+extendVarEnv :: Name () -> Plain Value -> VarEnv -> VarEnv -> VarEnv
+extendVarEnv argName arg argVarEnv (VarEnv oldVarEnv) =
+    VarEnv $ IntMap.insert (unUnique $ nameUnique argName) (Closure argVarEnv arg) oldVarEnv
+
+-- | Look up a variable name in the environment.
+lookupVarName :: Name () -> CekM Closure
+lookupVarName varName = do
+    VarEnv varEnv <- getVarEnv
+    case IntMap.lookup (unUnique $ nameUnique varName) varEnv of
+        Nothing   -> throwError $ MachineException OpenTermEvaluatedMachineError (Var () varName)
+        Just clos -> pure clos
+
+-- | Look up a 'DynamicBuiltinName' in the environment.
+lookupDynamicBuiltinName :: DynamicBuiltinName -> CekM (Maybe DynamicBuiltinNameMeaning)
+lookupDynamicBuiltinName dynName =
+    asks $ Map.lookup dynName . unDynamicBuiltinNameMeanings . _cekEnvDbnms
```

Comment:
Oh I see, we're first making a `ConstAppErr` which then becomes an `MachineError`...

Reply:
> Why not do the same as in the previous function and throw an error here if it's missing?

Because the set of allowed built-ins is controlled by the user, so a missing dynamic built-in is not an internal error.

Reply:
Ah, sorry, didn't look at the module header. Yeah, because that's not an exception, but causes `ConstAppFailure` which results in `EvaluationFailure` -- not `MachineError`. `EvaluationFailure` is not an error, it's a result of evaluation (the out of gas situation causes `EvaluateFailure` as well). I explained this design decision in the ticket:

> the machines deliberately do not return any errors -- just `Failure` -- and hence for now evaluation can return `Failure` for scope resolution failures as well.

It's fine for now, but once we change the approach we use for handling dynamic built-ins, there will be something better than that.

Reply:
However

> Because the set of allowed built-ins is controlled by the user, so a missing dynamic built-in is not an internal error.

is related here. We can't simply return an error, because it's not an internal one and all errors that the CEK machine returns are internal ones. Hence I'm returning a `Failure`.

Reply:
Okay - perhaps we could document some of these design decisions in the code too?

Reply:
I realized that since I'm now aware that we do not have all uniques equal to each other after the parsing step, the CEK machine should be designed the way you prefer. I opened a ticket: CGP-386. I do not want to make any radical changes in this PR, but since all the errors are now not internal, it's perfectly fine to add the `UnknownDynamicBuiltinNameError` to them which I've just done.